File: src/main/java/coffee/amo/interactionkey/net/ServerboundInteractionPacket.java
package coffee.amo.interactionkey.net;

import coffee.amo.interactionkey.api.event.InteractionKeyPressedEvent;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.network.NetworkEvent;

import java.util.function.Supplier;

public class ServerboundInteractionPacket {
    public ServerboundInteractionPacket() {
    }

    public static void encode(ServerboundInteractionPacket packet, FriendlyByteBuf buffer) {
    }

    public static ServerboundInteractionPacket decode(FriendlyByteBuf buffer) {
        return new ServerboundInteractionPacket();
    }

    public static void handle(ServerboundInteractionPacket packet, Supplier<NetworkEvent.Context> ctx) {
        ctx.get().enqueueWork(() -> {
            MinecraftForge.EVENT_BUS.register(new InteractionKeyPressedEvent(ctx.get().getSender()));
        });
        ctx.get().setPacketHandled(true);
    }


}

File: src/main/kotlin/coffee/amo/interactionkey/net/ServerboundInteractionPacket.kt
package coffee.amo.interactionkey.net

import coffee.amo.interactionkey.api.event.InteractionKeyPressedEvent
import net.minecraft.core.BlockPos
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Player
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.network.NetworkEvent
import java.util.function.Supplier

class ServerboundInteractionPacket(val blockPos: BlockPos, val entityID: Int) {
    companion object{
        fun encode(packet: ServerboundInteractionPacket?, buffer: FriendlyByteBuf?) {
            buffer!!.writeBlockPos(packet!!.blockPos)
            buffer.writeVarInt(packet.entityID)
        }

        fun decode(buffer: FriendlyByteBuf?): ServerboundInteractionPacket {
            return ServerboundInteractionPacket(buffer!!.readBlockPos(), buffer.readVarInt())
        }

        fun handle(packet: ServerboundInteractionPacket?, ctx: Supplier<NetworkEvent.Context?>?) {
            ctx!!.get()!!.enqueueWork {
                val player = ctx.get()!!.sender as Player
                val blockPos = packet!!.blockPos
                val entityID = packet.entityID
                MinecraftForge.EVENT_BUS.post(InteractionKeyPressedEvent(player, blockPos, entityID))
                println("Received interaction packet from ${player.name} at $blockPos with entityID $entityID")
            }
            ctx.get()!!.packetHandled = true
        }
    }


}